### PR TITLE
HBASE-22504 Optimize the MultiByteBuff#get(ByteBuffer, offset, len)

### PR DESCRIPTION
**Pull request by openinx**

### CODE REVIEW

**Comment by Apache-HBase** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 220 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | hbaseanti | 0 |  Patch does not have any anti-patterns. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 1 new or modified test files. |
||| _ HBASE-21879 Compile Tests _ |
| +1 | mvninstall | 277 | HBASE-21879 passed |
| +1 | compile | 20 | HBASE-21879 passed |
| +1 | checkstyle | 22 | HBASE-21879 passed |
| +1 | shadedjars | 253 | branch has no errors when building our shaded downstream artifacts. |
| +1 | findbugs | 37 | HBASE-21879 passed |
| +1 | javadoc | 18 | HBASE-21879 passed |
||| _ Patch Compile Tests _ |
| +1 | mvninstall | 239 | the patch passed |
| +1 | compile | 21 | the patch passed |
| +1 | javac | 21 | the patch passed |
| -1 | checkstyle | 24 | hbase-common: The patch generated 1 new + 73 unchanged - 1 fixed = 74 total (was 74) |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedjars | 278 | patch has no errors when building our shaded downstream artifacts. |
| +1 | hadoopcheck | 573 | Patch does not cause any errors with Hadoop 2.7.4 or 3.0.0. |
| +1 | findbugs | 46 | the patch passed |
| +1 | javadoc | 19 | the patch passed |
||| _ Other Tests _ |
| +1 | unit | 165 | hbase-common in the patch passed. |
| +1 | asflicense | 11 | The patch does not generate ASF License warnings. |
| | | 2299 | |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=17.05.0-ce Server=17.05.0-ce base: https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-273/1/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/273 |
| Optional Tests |  dupname  asflicense  javac  javadoc  unit  findbugs  shadedjars  hadoopcheck  hbaseanti  checkstyle  compile  |
| uname | Linux 59cf6fc09c54 4.4.0-139-generic #165-Ubuntu SMP Wed Oct 24 10:58:50 UTC 2018 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | /testptch/patchprocess/precommit/personality/provided.sh |
| git revision | HBASE-21879 / 68c51293ab |
| maven | version: Apache Maven 3.5.4 (1edded0938998edf8bf061f1ceb3cfdeccf443fe; 2018-06-17T18:33:14Z) |
| Default Java | 1.8.0_181 |
| findbugs | v3.1.11 |
| checkstyle | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-273/1/artifact/out/diff-checkstyle-hbase-common.txt |
|  Test Results | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-273/1/testReport/ |
| Max. process+thread count | 363 (vs. ulimit of 10000) |
| modules | C: hbase-common U: hbase-common |
| Console output | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-273/1/console |
| Powered by | Apache Yetus 0.9.0 http://yetus.apache.org |


This message was automatically generated.

File: hbase-common/src/main/java/org/apache/hadoop/hbase/nio/MultiByteBuff.java [hunk 1]
```diff
@@ -1020,24 +1019,28 @@ public void asSubByteBuffer(int offset, int length, ObjectIntPair<ByteBuffer> pa
...
+  public void get(ByteBuffer out, int sourceOffset, int length) {
     checkRefCount();
-      // Not used from real read path actually. So not going with
-      // optimization
-    for (int i = 0; i < length; ++i) {
-      out.put(this.get(sourceOffset + i));
+    int itemIndex = getItemIndex(sourceOffset);
+    ByteBuffer in = this.items[itemIndex];
+    sourceOffset = sourceOffset - this.itemBeginPos[itemIndex];
+    while (length > 0) {
```
Comment (ramkrish86) [inline hbase-common/src/main/java/org/apache/hadoop/hbase/nio/MultiByteBuff.java]: LGTM.

Comment (openinx) [reply hbase-common/src/main/java/org/apache/hadoop/hbase/nio/MultiByteBuff.java]: Please click the [Review changes], and select the [approve],  then I can merge this PR :-)  Thanks.

**Comment by openinx** [comment]
Addressed the checkstyle issue from HBase-QA

**Comment by Apache-HBase** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 278 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | hbaseanti | 0 |  Patch does not have any anti-patterns. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 1 new or modified test files. |
||| _ HBASE-21879 Compile Tests _ |
| +1 | mvninstall | 277 | HBASE-21879 passed |
| +1 | compile | 20 | HBASE-21879 passed |
| +1 | checkstyle | 23 | HBASE-21879 passed |
| +1 | shadedjars | 288 | branch has no errors when building our shaded downstream artifacts. |
| +1 | findbugs | 50 | HBASE-21879 passed |
| +1 | javadoc | 22 | HBASE-21879 passed |
||| _ Patch Compile Tests _ |
| +1 | mvninstall | 306 | the patch passed |
| +1 | compile | 23 | the patch passed |
| +1 | javac | 23 | the patch passed |
| +1 | checkstyle | 23 | hbase-common: The patch generated 0 new + 73 unchanged - 1 fixed = 73 total (was 74) |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedjars | 277 | patch has no errors when building our shaded downstream artifacts. |
| +1 | hadoopcheck | 558 | Patch does not cause any errors with Hadoop 2.7.4 or 3.0.0. |
| +1 | findbugs | 47 | the patch passed |
| +1 | javadoc | 19 | the patch passed |
||| _ Other Tests _ |
| +1 | unit | 162 | hbase-common in the patch passed. |
| +1 | asflicense | 10 | The patch does not generate ASF License warnings. |
| | | 2459 | |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=17.05.0-ce Server=17.05.0-ce base: https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-273/2/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/273 |
| Optional Tests |  dupname  asflicense  javac  javadoc  unit  findbugs  shadedjars  hadoopcheck  hbaseanti  checkstyle  compile  |
| uname | Linux 9bea849cd0e2 4.4.0-137-generic #163-Ubuntu SMP Mon Sep 24 13:14:43 UTC 2018 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | /testptch/patchprocess/precommit/personality/provided.sh |
| git revision | HBASE-21879 / 68c51293ab |
| maven | version: Apache Maven 3.5.4 (1edded0938998edf8bf061f1ceb3cfdeccf443fe; 2018-06-17T18:33:14Z) |
| Default Java | 1.8.0_181 |
| findbugs | v3.1.11 |
|  Test Results | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-273/2/testReport/ |
| Max. process+thread count | 343 (vs. ulimit of 10000) |
| modules | C: hbase-common U: hbase-common |
| Console output | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-273/2/console |
| Powered by | Apache Yetus 0.9.0 http://yetus.apache.org |


This message was automatically generated.

**Pull request merged**